sympy/core/function.py

```diff
@@ -1298,29 +1298,11 @@ def _remove_derived_once(cls, v):
 
     @classmethod
     def _sort_variable_count(cls, varcounts):
-        """Like ``_sort_variables``, but acting on variable-count pairs.
-
-        Examples
-        ========
-
-        >>> from sympy import Derivative, Function, symbols
-        >>> vsort = Derivative._sort_variable_count
-        >>> x, y, z = symbols('x y z')
-        >>> f, g, h = symbols('f g h', cls=Function)
-
-        >>> vsort([(x, 1), (y, 2), (z, 1)])
-        [(x, 1), (y, 2), (z, 1)]
-
-        >>> vsort([(z, 1), (y, 1), (x, 1), (h(x), 1), (g(x), 1), (f(x), 1)])
-        [(x, 1), (y, 1), (z, 1), (f(x), 1), (g(x), 1), (h(x), 1)]
         """
-        d = dict(varcounts)
-        varsorted = cls._sort_variables([i for i, j in varcounts])
-        return [Tuple(var, d[var]) for var in varsorted]
+        Sort (variable, count) pairs by variable, but disallow sorting of non-symbols.
```

Comment:
Is this right? It looks like the non-symbols *are* sorted: `sorted(non_symbol_part,...`

It looks like the current approach sorts the wrt variables batches which are contiguous wrt the variable type: symbol or non-symbol. This is conservative. It could actually sort non-symbols with symbols provided that there were no free symbols in common, e.g. x, y, f(z) could appear in arbitrary sorted order but x, y, f(x) could not appear as f(x), x, y.

Reply:
Yeah I agree. This is the behavior that was there before (this PR just fixes the behavior wrt the counts). @Upabjojr will need to comment on what the correct behavior should be. Is the idea just that we can't sort x and f(x) because f(x) has x as a free symbol or are there other reasons to not sort symbols and non-symbols?

Reply:
inline says 
```
# If we wanted to evaluate, we sort the variables into standard
# order for later comparisons. 
```

So better canonicalization will be had if any symbols, appearing after functions, that are not in the free symbols of the function, appear before the functions: `x, f(y), x, y -> x, x, f(y), y` (except with counts included).

But that is a separate issue. You've addressed what you set out to fix.